src/stan/model/indexing/assign_varmat.hpp

```diff
@@ -147,22 +245,25 @@ inline void assign(Vec1&& x, const Vec2& y, const char* name,
  * @throw std::out_of_range If either of the indices are out of bounds.
  */
 template <typename Mat, typename U, require_var_dense_dynamic_t<Mat>* = nullptr,
-          require_var_t<U>* = nullptr,
-          require_floating_point_t<value_type_t<U>>* = nullptr>
+          require_stan_scalar_t<U>* = nullptr>
 inline void assign(Mat&& x, const U& y, const char* name, index_uni row_idx,
                    index_uni col_idx) {
   stan::math::check_range("matrix[uni,uni] assign", name, x.rows(), row_idx.n_);
   stan::math::check_range("matrix[uni,uni] assign", name, x.cols(), col_idx.n_);
   const int row_idx_val = row_idx.n_ - 1;
   const int col_idx_val = col_idx.n_ - 1;
   double prev_val = x.val().coeffRef(row_idx_val, col_idx_val);
-  x.vi_->val_.coeffRef(row_idx_val, col_idx_val) = y.val();
-  stan::math::reverse_pass_callback(
-      [x, y, row_idx_val, col_idx_val, prev_val]() mutable {
-        x.vi_->val_.coeffRef(row_idx_val, col_idx_val) = prev_val;
-        y.adj() += x.adj().coeffRef(row_idx_val, col_idx_val);
-        x.adj().coeffRef(row_idx_val, col_idx_val) = 0.0;
-      });
+  x.vi_->val_.coeffRef(row_idx_val, col_idx_val) = stan::math::value_of(y);
+  stan::math::reverse_pass_callback([x, y, row_idx_val, col_idx_val,
+                                     prev_val]() mutable {
+    x.vi_->val_.coeffRef(row_idx_val, col_idx_val) = prev_val;
+    using stan::math::forward_as;
+    using stan::math::var;
```

Comment:
[optional] I am not a fan of using declarations for functions and types used only once.